Add animation slices for individual animations

Juankz opened the pull request:
The clips method only worked if the imported model had an animation named "default". This PR allows clips to be created no matter what the one-track animation name is.

This closes my other PR #36709.

*Bugsquad edit:*
- Closes https://github.com/godotengine/godot/issues/25599.
- Closes #64486.

fire commented on editor/import/resource_importer_scene.cpp:
```diff
@@ -648,11 +648,14 @@ void ResourceImporterScene::_create_clips(Node *scene, const Array &p_clips, boo
+	List<StringName> default_animations;
+	anim->get_animation_list(&default_animations);
+
+	if (default_animations.size() <= 0) {
```
> if (!default_animations.size()) {

The range of size() is 0 to the maximum int

Juankz replied:
Thank you

Juankz commented:
I attach a sample project [anim-clips.zip](https://github.com/godotengine/godot/files/4907039/anim-clips.zip)

noidexe commented:
Is this merged on stable? Any ETAs?

akien-mga commented:
Well this needs a review from @reduz who is more familiar with animation workflow.

For what it's worth, while I see value in allowing the use of other default animations than one named "default", I don't like that it arbitrarily picks the first one. It's a valid workflow to actually have an animation named "default", which may not be the first one, but which you want to be used as default.

noidexe commented:
>  I don't like that it arbitrarily picks the first one.
Me neither but I've noticed more than one animator struggling with this and it seems that when using clips having one animation is the most common use case, while having it named "default" isn't, but rather the default named assigned by the 3dcg sofware. 

The engine doesn't report any error when you try to clip with a different name, it just creates the wrong animations, so it becomes a RTFM moment for the integrator/animator.

If you a dozen of purchased assets it means re exporting every one of those which might even involve getting the software and learning how to do it in what would otherwise work by default. 

I think a way it can work without breaking the workflow of those who are following the docs would be the following:
- If there is an animation named 'default' use that for clipping
- Otherwise use the first one as a fallback rather than failing

That would work well for people explicitly using "default" as well as for assets with only one animation with whatever name.

fire commented:
> The clips method only worked if the imported model had an animation named "default".

I can sense the usefulness of using the first track if it isn't named "default". However, the usage of "default" when there there are two tracks needs to be designed.

reduz commented:
Is this still the case in 4.0?

Juankz commented:
I haven't been able to use clips successfully in master branch with the test model I attached above. It keeps being imported as one animation track

YuriSizov commented:
@godotengine/animation Can you confirm if this is still needed?

TokageItLab commented:
I think it is good, but @fire is the best judge of this feature.

fire commented:
Common workflow is to use one track and then split. Very wanted

YuriSizov commented:
Right, then this needs a rebase and a review.

reduz commented:
This makes sense, would be good to salvage.

Juankz commented:
I'll take another look into it

Juankz committed:
Add animation slices for individual animations

This improves the workflow for animations in a single timeline.
The users are no longer forced to slice one animation named "default".
Instead users can choose which animation(s) to break and how.

Changes:

- Remove slicing options from the animation player import menu
- Add such options to the animation import menu
- Rename clips to slices wherever was left

Juankz committed:
Connect signal animation_removed to its proper method

Juankz commented:
All right. Here's what I did:

- Remove slicing options from the animation player import menu
- Add such options to the animation import menu
- Rename clips to slices wherever was left to be done. This was actually breaking the current workflow.

So the import menu looks like this.
![Screenshot_20221010_191153](https://user-images.githubusercontent.com/18577412/194970827-9fcd8759-6a37-4d90-82a1-fae9422010e6.png)

Now instead of using "default" or picking arbitrarily the first one, the user can decide which animation to slice.

Here for example there are two timelines that needed slicing: One the body movement and another for facial animations. Such workflow is now possible with this PR.

![6344baf567c46316572123](https://user-images.githubusercontent.com/18577412/194972861-2d685e3e-657e-45ed-a4c4-e0e45a46f01c.gif)

Here is my demo project. Let me know what you think:
[AnimationSlicesDemo.tar.gz](https://github.com/godotengine/godot/files/9750041/AnimationSlicesDemo.tar.gz)

fire reviewed (approved):
https://user-images.githubusercontent.com/32321/195012146-2475ba8d-6b0b-466f-97d2-64c376479083.mp4

The animation slice feature works like this:

1. for each animation ".anim" we have slices
2. You can set the slice to 0, the clips in the files or some number.
3. The point of this feature is for each "clip" you can remix the slices order and naming based on the frame count.
4. When the number is less, less clips are output
5. When the number is equal, there is no splitting
6. When it is more, clips are invented.
7. Might want to have a reset to default number of slices.

This is a feature review. Need to review the code.

fire commented on editor/import/resource_importer_scene.cpp:
```diff
@@ -1408,144 +1414,138 @@ Ref<Animation> ResourceImporterScene::_save_animation_to_file(Ref<Animation> ani
-		bool save_to_file = p_clips[i + 4];
-		String save_to_path = p_clips[i + 5];
-		bool keep_current = p_clips[i + 6];
+	for (int i = 0; i < p_slices.size(); i += 7) {
```
What does 7 mean, can you make it a named constant and the same with the ones below?

Juankz replied:
I didn't touch most of the slicing logic, I just updated the variable names. I see it's 7 because each slice contains 7 import options. But it was like that before. I can refactor it once I get some more free time. Also I would like to see if there are more comments on this PR.

akien-mga commented:
Thanks!
